orderer/common/cluster/commauth.go

```diff
@@ -0,0 +1,331 @@
+func (ac *AuthCommMgr) updateStubInMapping(channel string, mapping MemberMapping, node NodeAddress) {
+	stub := mapping.ByID(node.ID)
+	if stub == nil {
+		ac.Logger.Infof("Allocating a new stub for node %v with endpoint %v for channel %s", node.ID, node.Endpoint, channel)
+		stub = &Stub{}
+	}
+
+	// Overwrite the stub Node data with the new data
+	stub.RemoteNode.NodeAddress = node
+
+	// Put the stub into the mapping
+	mapping.Put(stub)
+
+	// Check if the stub needs activation.
+	if stub.Active() {
+		return
+	}
+
+	// Activate the stub
+	stub.Activate(ac.createRemoteContext(stub, channel))
+}
+
+func (ac *AuthCommMgr) createRemoteContext(stub *Stub, channel string) func() (*RemoteContext, error) {
+	return func() (*RemoteContext, error) {
+		ac.Logger.Debugf("Connecting to node: %v for channel: %v", stub.RemoteNode.NodeAddress, channel)
+
+		conn, err := ac.Connections.Connect(stub.Endpoint)
+		if err != nil {
+			ac.Logger.Warningf("Unable to obtain connection to %d(%s) (channel %s): %v", stub.ID, stub.Endpoint, channel, err)
+			return nil, err
+		}
+
+		probeConnection := func(conn *grpc.ClientConn) error {
+			connState := conn.GetState()
+			if connState == connectivity.Connecting {
+				return errors.Errorf("connection to %d(%s) is in state %s", stub.ID, stub.Endpoint, connState)
+			}
+			return nil
+		}
+
+		clusterClient := orderer.NewClusterNodeServiceClient(conn)
+		getStepClientStream := func(ctx context.Context) (StepClientStream, error) {
+			stream, err := clusterClient.Step(ctx)
+			if err != nil {
+				return nil, err
+			}
+			stepClientStream := &NodeClientStream{
+				StepClient: stream,
+			}
+			return stepClientStream, nil
+		}
+
+		workerCountReporter := workerCountReporter{
+			channel: channel,
+		}
+
+		rc := &RemoteContext{
+			Metrics:             ac.Metrics,
+			workerCountReporter: workerCountReporter,
+			Channel:             channel,
+			SendBuffSize:        ac.SendBufferSize,
+			endpoint:            stub.Endpoint,
+			Logger:              ac.Logger,
+			ProbeConn:           probeConnection,
+			conn:                conn,
+			GetStreamFunc:       getStepClientStream,
+			// Client:              clusterClient,
+			shutdownSignal:    ac.shutdownSignal,
+			SourceNodeID:      ac.NodeID,
+			DestinationNodeID: stub.ID,
+			Signer:            ac.Signer,
+		}
+		return rc, nil
+	}
+}
+
+type NodeClientStream struct {
+	StepClient orderer.ClusterNodeService_StepClient
+}
+
+func (cs *NodeClientStream) Send(request *orderer.StepRequest) error {
+	stepRequest, cerr := BuildStepRequest(request)
+	if cerr != nil {
+		return cerr
+	}
+	return cs.StepClient.Send(stepRequest)
+}
+
+func (cs *NodeClientStream) Recv() (*orderer.StepResponse, error) {
+	nodeResponse, err := cs.StepClient.Recv()
+	if err != nil {
+		return nil, err
+	}
+	return BuildStepRespone(nodeResponse)
+}
+
+func (cs *NodeClientStream) Auth(version uint32, fromId uint64, toId uint64, channel string, signer identity.Signer) error {
```

Comment:
Are you sure you need to pass in all these arguments? Can't you initialize them as fields?

Reply:
Its not required. Updated the Auth interface.